rdmd.d

```diff
@@ -674,42 +738,211 @@ private string[string] getDependencies(string rootModule, string workDir,
         if (depsT > SysTime.min)
         {
             // See if the deps file is still in good shape
-            auto deps = readDepsFile();
-            auto allDeps = chain(rootModule.only, deps.byKey).array;
+            deps = readDepsFile();
+            auto allDeps = deps.byKey
+                .map!(node => node.file)
+                .filter!(f => f != "")
+                .array;
             bool mustRebuildDeps = allDeps.anyNewerThan(depsT);
             if (!mustRebuildDeps)
             {
                 // Cool, we're in good shape
-                return deps;
+                generateDepsFile = false;
+            }
+        }
+    }
+
+    if (generateDepsFile)
+    {
+        // Don't pass -lib to the compiler here.
+        compilerFlags = compilerFlags.filter!(flag => flag != "-lib").array();
+
+        // Collect dependencies
+        auto depsGetter =
+            [ compiler ] ~ compilerFlags ~
+            ["-deps", "-o-", rootModule];
+
+        scope(failure)
+        {
+            // Delete the deps file on failure, we don't want to be fooled
+            // by it next time we try
+            collectException(std.file.remove(depsFilename));
```

Comment:
The more robust approach here is to generate deps in a different filename, e.g. `depsFilename ~ ".truncated"`, and upon success rename it into `depsFilename`.